src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Signed.Default.cs

```diff
@@ -0,0 +1,64 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers.Text
+{
+    /// <summary>
+    /// Methods to format common data types as Utf8 strings.
+    /// </summary>
+    public static partial class Utf8Formatter
+    {
+        private static bool TryFormatInt64Default(long value, Span<byte> buffer, out int bytesWritten)
+        {
+            ref byte utf8Bytes = ref buffer.DangerousGetPinnableReference();
```

Comment:
I do not think it is a good idea to be using DangerousGetPinnableReference and `ref` math in methods like these. It leads to hard to read code. Instead, you should pin the string using `fixed` and write the code using pointers old-fashioned way.

Reply:
I can make this change. However, the previous implementation already did this.

The same issue occurs elsewhere:
https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Signed.D.cs#L30

I will open an issue to address all of them outside this PR.

Reply:
@jkotas why not jut use the span like an array? I thought with the JIT changes we could get pretty close without resorting to pointers.

Reply:
Yes, I agree. This code is not written the way I think it should be written. 

I have commented on this here as well: https://github.com/dotnet/corefx/pull/25078/files#r149250107 . I would much rather take 1-2% regression here and have a nice safe readable code than have piles of unsafe code that are large security liability.

Reply:
> I would much rather take 1-2% regression here and have a nice safe readable code than have piles of unsafe code that are large security liability.

I observed a larger regression (5%+) for large numbers (median of 3 runs).
![image](https://user-images.githubusercontent.com/6527137/33297722-75f7b48e-d397-11e7-8b5c-7ea601395725.png)

Reply:
Passing the span and using the indexer is 5-10% slower than using Unsafe APIs on ref byte.
Should I make the change to use [this](https://github.com/dotnet/corefx/pull/25424/files#diff-49c895af5389520da847da981958896bR59) over `ref byte`?
```C#
public static void WriteDigits(ulong value, int digitCount, Span<byte> buffer, int index)
```


Test   Name | Unsafe | Indexer | Ratio
-- | -- | -- | --
FormatterUInt64(value: 0) | 11.796 | 11.657 | 1.012
FormatterUInt64(value: 12837467) | 23.475 | 21.518 | 1.091
FormatterUInt64(value:   18446744073709551615) | 43.335 | 39.978 | 1.084

Reply:
Have you tried to write it like this? It should reduce the register pressure for the loop quite a bit:

```
static void WriteDigits(ulong value, Span<byte> buffer)
{
    ulong left = value;

    for (int i = buffer.Length - 1; i >= 1; i--)
    {
        left = DivMod(left, 10, out ulong num);
        buffer[i] = (byte)('0' + num);
    }

    buffer[0] = (byte)('0' + left);
}
```

You will likely find more opportunities for optimizations once you look at the generated code. In particular, I am not sure what the extensive use of AggresiveInlining does to the code. I would be nice to structure the code differently so that you do not have to use it as much for relatively large methods.

Reply:
> Have you tried to write it like this?

```C#
// A) 5-10% faster compared to B:
FormattingHelpers.WriteDigits((ulong)value, digits, ref buffer.DangerousGetPinnableReference());
// B)
FormattingHelpers.WriteDigits((ulong)value, buffer.Slice(0, digits));
```

Aggressive Inlining helps in A) but hurts in B).